### PR TITLE
ci: improve parallelisation

### PR DESCRIPTION
**Pull request by lavocatt**
Run more than once the same arch/distro in parallel. Each instance will work on the other half of the test cases to run. All tests cases will be executed but twice as fast.

### CODE REVIEW

File: tools/osbuild-image-test [hunk 1]
```diff
@@ -402,7 +402,21 @@ def run_tests_cases(args):
...
+    #
+    # If the user specified that more than one instances are sharing the tests
+    # set, then figure out which tests to execute based on the instance number
+    # over the total number of instances.
+    bucket_number: int = args.instance_number - 1
+    bucket_size: int = math.ceil(len(tests) / args.nb_instances)
+
+    start_index: int = bucket_number * bucket_size
+    end_index: int = (bucket_number + 1) * bucket_size
+    end_index = end_index if end_index <= len(tests) else len(tests)
```
Comment (ondrejbudai) [inline tools/osbuild-image-test]:
```suggestion
    end_index = min(end_index, len(tests))
```

File: tools/osbuild-image-test [hunk 2]
```diff
@@ -402,7 +402,21 @@ def run_tests_cases(args):
...
+
+    for test in tests[start_index: end_index]:
```
Comment (ondrejbudai) [inline tools/osbuild-image-test]: There is usually no whitespace inside the Python's slicing operator:

```suggestion
    for test in tests[start_index:end_index]:
```

File: test/cases/gen_db [hunk 3]
```diff
@@ -21,14 +21,23 @@ distro = distro.replace('"', '')
...
+           "--generator-mode"]
+
+pa_exec = sys.argv[1] if len(sys.argv) > 1 else None
+if pa_exec:
+    total = int(pa_exec.split("/")[1])
+    part = int(pa_exec.split("/")[0])
+    command.append(f"--instance-number={part}")
+    command.append(f"--nb-instances={total}")
+
+print(" ".join(command), flush=True)
```
Comment (ondrejbudai) [inline test/cases/gen_db:38]: A stray print? Or do you want to leave this in?

File: test/cases/manifest_tests [hunk 4]
```diff
@@ -20,13 +20,23 @@ distro = f"{config.get('DEFAULT', 'ID')}-{config.get('DEFAULT', 'VERSION_ID')}"
...
+           f"--distro={distro}"]
+
+pa_exec = sys.argv[1] if len(sys.argv) > 1 else None
+if pa_exec:
+    total = int(pa_exec.split("/")[1])
+    part = int(pa_exec.split("/")[0])
+    command.append(f"--instance-number={part}")
+    command.append(f"--nb-instances={total}")
+
+print(" ".join(command), flush=True)
```
Comment (ondrejbudai) [inline test/cases/manifest_tests:38]: Same here.

File: tools/osbuild-image-test [hunk 5]
```diff
@@ -540,6 +554,18 @@ def main():
         default="manifest-db",
         help="path to the database folder"
     )
+    parser.add_argument(
+        "--instance-number",
+        type=int,
+        default=1,
+        help="if several identical instances are running specify it's number here"
```
Comment (ondrejbudai) [inline tools/osbuild-image-test]:
```suggestion
        help="if several identical instances are running, specify its number here"
```

File: tools/osbuild-image-test [hunk 6]
```diff
@@ -540,6 +554,18 @@ def main():
...
+    )
+    parser.add_argument(
+        "--nb-instances",
```
Comment (ondrejbudai) [inline tools/osbuild-image-test]: I wonder: What does `nb` mean? :)

**Review by ondrejbudai — CHANGES_REQUESTED** [review]
Simple, yet very effective, I like it! Some comments inline, nothing major, I believe we should be able to merge this soon. :)

File: tools/osbuild-image-test (hunk 6 above)
Comment (lavocatt) [reply tools/osbuild-image-test]: Ahah, that might be a French habit ? lol, It's for the total number of instances, maybe renaming it `--total-number-of-instances` might be better ? What do you prefer ?

File: test/cases/gen_db (hunk 3 above)
Comment (lavocatt) [reply test/cases/gen_db:38]: I'd rather leave that in, a bit of debug, might be always useful. You know just to know how many instances are running. Also I think, I'm gonna add a bit more logs in the main program, to make sure we always run all the test cases.

File: tools/osbuild-image-test (hunk 2 above)
Comment (lavocatt) [reply tools/osbuild-image-test]: This Autopep8 is great, except when it doesn't run on non .py files that python ones. Good catch thanks!!

File: tools/osbuild-image-test (hunk 1 above)
Comment (lavocatt) [reply tools/osbuild-image-test]: This is better!

**Comment by lavocatt** [comment]
> Simple, yet very effective, I like it! Some comments inline, nothing major, I believe we should be able to merge this soon. :)

Thanks!! I'm gonna make another round of improvement in the parallelization. Some instances are taking 2 hours to complete, I'm gonna target these ones to split them in 4 pieces instead of 2 to speed up everything.

File: tools/osbuild-image-test (hunk 2 above)
Comment (lavocatt) [reply tools/osbuild-image-test]: Interesting, after running Autopep8 on the file, it doesn't care about this whitespace. I'm removing it manually.

**Comment by lavocatt** [comment]
Before we can accept this PR, I will need to run a fake update against it to see if the workflow is still valid with parallelization.

File: tools/osbuild-image-test (hunk 6 above)
Comment (ondrejbudai) [reply tools/osbuild-image-test]: `--num-instances` is what I would use in English. :)

**Review by ondrejbudai — APPROVED** [review]

File: tools/osbuild-image-test (hunk 6 above)
Comment (lavocatt) [reply tools/osbuild-image-test]: Do you want me to change it ? Or is  `--total-number-of-instances` good enough ?

Comment (ondrejbudai) [reply tools/osbuild-image-test]: Well, it's quite long, but also descriptive. I don't mind either, feel to merge, or rename, I have no strong preference. :)

**Pull request merged**